pbrd/pbr_vty.c

```diff
@@ -153,325 +128,949 @@ DEFUN_NOSH(no_pbr_map, no_pbr_map_cmd, "no pbr-map PBRMAP [seq (1-700)]",
+	const char *fmsg = NULL;
+
+	if (!pbrms)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (no) {
+		if (!CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_SRC_IP))
+			return CMD_SUCCESS;
+		prefix_free(&pbrms->src);
+		UNSET_FLAG(pbrms->filter_bm, PBR_FILTER_SRC_IP);
+		goto check;
+	}
+
+	assert(prefix);
+	if (!pbr_family_consistent(pbrms, prefix->family, PBR_FILTER_SRC_IP, 0,
+				   &fmsg)) {
+		vty_out(vty, "Address family mismatch (%s)\n", fmsg);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+	pbrms->family = prefix->family;
+
+	if (CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_SRC_IP)) {
+		if (prefix_same(pbrms->src, prefix))
+			return CMD_SUCCESS;
+	} else
+		pbrms->src = prefix_new();
+
+	prefix_copy(pbrms->src, prefix);
+	SET_FLAG(pbrms->filter_bm, PBR_FILTER_SRC_IP);
+
+check:
+	pbr_map_check(pbrms, true);
+
 	return CMD_SUCCESS;
 }
 
-DEFPY(pbr_map_match_src, pbr_map_match_src_cmd,
-	"[no] match src-ip <A.B.C.D/M|X:X::X:X/M>$prefix",
+/* clang-format off */
+DEFPY  (pbr_map_match_dst,
+	pbr_map_match_dst_cmd,
+	"[no] match dst-ip ![<A.B.C.D/M|X:X::X:X/M>$prefix]",
 	NO_STR
 	"Match the rest of the command\n"
-	"Choose the src ip or ipv6 prefix to use\n"
+	"Choose the dst ip or ipv6 prefix to use\n"
 	"v4 Prefix\n"
 	"v6 Prefix\n")
 {
+	/* clang-format on */
+	struct pbr_map_sequence *pbrms = VTY_GET_CONTEXT(pbr_map_sequence);
+	const char *fmsg = NULL;
+
+	if (!pbrms)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (no) {
+		if (!CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_DST_IP))
+			return CMD_SUCCESS;
+		prefix_free(&pbrms->dst);
+		UNSET_FLAG(pbrms->filter_bm, PBR_FILTER_DST_IP);
+		goto check;
+	}
+
+	assert(prefix);
+	if (!pbr_family_consistent(pbrms, prefix->family, PBR_FILTER_DST_IP, 0,
+				   &fmsg)) {
+		vty_out(vty, "Address family mismatch (%s)\n", fmsg);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+	pbrms->family = prefix->family;
+
+	if (CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_DST_IP)) {
+		if (prefix_same(pbrms->dst, prefix))
+			return CMD_SUCCESS;
+	} else
+		pbrms->dst = prefix_new();
+
+	prefix_copy(pbrms->dst, prefix);
+	SET_FLAG(pbrms->filter_bm, PBR_FILTER_DST_IP);
+
+check:
+	pbr_map_check(pbrms, true);
+
+	return CMD_SUCCESS;
+}
+
+/* clang-format off */
+DEFPY  (pbr_map_match_ip_proto,
+	pbr_map_match_ip_proto_cmd,
+	"[no] match ip-protocol ![PROTO$ip_proto]",
+	NO_STR
+	"Match the rest of the command\n"
+	"Choose an ip-protocol\n"
+	"Protocol name\n")
+{
+	/* clang-format on */
 	struct pbr_map_sequence *pbrms = VTY_GET_CONTEXT(pbr_map_sequence);
+	struct protoent *p;
 
 	if (!pbrms)
 		return CMD_WARNING_CONFIG_FAILED;
 
-	if (pbrms->dst && prefix->family != pbrms->dst->family) {
-		vty_out(vty, "Cannot mismatch families within match src/dst\n");
+	if (no) {
+		if (!CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_IP_PROTOCOL))
+			return CMD_SUCCESS;
+		UNSET_FLAG(pbrms->filter_bm, PBR_FILTER_IP_PROTOCOL);
+		goto check;
+	}
+
+	if (!ip_proto || !((p = getprotobyname(ip_proto)))) {
```

Comment:
Can we have `p = ...` check outside this scope for readability?

Reply:
Agreed - reworked logic a bit.